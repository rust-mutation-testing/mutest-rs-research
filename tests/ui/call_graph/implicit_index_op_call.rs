//@ print-call-graph
//@ print-targets
//@ stdout
//@ stderr: empty

use std::ops::{Index, IndexMut};

struct ImplsIndex;

fn index_impl() {}
impl Index<usize> for ImplsIndex {
    type Output = Self;

    fn index(&self, _index: usize) -> &Self::Output {
        index_impl();
        self
    }
}

fn index_mut_impl() {}
impl IndexMut<usize> for ImplsIndex {
    fn index_mut(&mut self, _index: usize) -> &mut Self::Output {
        index_mut_impl();
        self
    }
}

#[test]
fn test_implicit_index_call() {
    let _ = ImplsIndex[0];
}

#[test]
fn test_implicit_index_mut_call() {
    ImplsIndex[0] = ImplsIndex;
}
