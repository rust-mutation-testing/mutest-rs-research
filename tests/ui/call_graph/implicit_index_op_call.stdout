
@@@ call graph @@@

entry points:

test test_implicit_index_call
  -> <ImplsIndex as std::ops::Index<usize>>::index at tests/ui/call_graph/implicit_index_op_call.rs:14:5: 14:52 (#0)
test test_implicit_index_mut_call
  -> <ImplsIndex as std::ops::IndexMut<usize>>::index_mut at tests/ui/call_graph/implicit_index_op_call.rs:22:5: 22:64 (#0)

nested calls at distance 1:

<ImplsIndex as std::ops::Index<usize>>::index at tests/ui/call_graph/implicit_index_op_call.rs:14:5: 14:52 (#0)
  -> index_impl at tests/ui/call_graph/implicit_index_op_call.rs:10:1: 10:16 (#0)
<ImplsIndex as std::ops::IndexMut<usize>>::index_mut at tests/ui/call_graph/implicit_index_op_call.rs:22:5: 22:64 (#0)
  -> index_mut_impl at tests/ui/call_graph/implicit_index_op_call.rs:20:1: 20:20 (#0)

@@@ targets @@@

tests -(1)-> index_impl at tests/ui/call_graph/implicit_index_op_call.rs:10:1: 10:16 (#0)
  (1) test_implicit_index_call

tests -(0)-> <ImplsIndex as std::ops::Index<usize>>::index at tests/ui/call_graph/implicit_index_op_call.rs:14:5: 14:52 (#0)
  (0) test_implicit_index_call

tests -(1)-> index_mut_impl at tests/ui/call_graph/implicit_index_op_call.rs:20:1: 20:20 (#0)
  (1) test_implicit_index_mut_call

tests -(0)-> <ImplsIndex as std::ops::IndexMut<usize>>::index_mut at tests/ui/call_graph/implicit_index_op_call.rs:22:5: 22:64 (#0)
  (0) test_implicit_index_mut_call

targets: 4 total; 4 safe; 0 unsafe (0 tainted)
